File: go/pkg/genx/transformers/minimax_tts.go
package transformers

import (
	"context"
	"io"
	"strings"

	"github.com/haivivi/giztoy/go/pkg/genx"
	"github.com/haivivi/giztoy/go/pkg/minimax"
)

// MinimaxTTS is a TTS transformer using MiniMax text-to-speech API.
//
// Model: speech-02-hd (default)
//
// Input type: text/plain
// Output type: audio/* (audio/mpeg by default)
//
// EoS Handling:
//   - When receiving a text/plain EoS marker, finish synthesis, emit audio chunks, then emit audio/* EoS
//   - Non-text chunks are passed through unchanged
type MinimaxTTS struct {
	client     *minimax.Client
	model      string
	voiceID    string
	speed      float64
	vol        float64
	pitch      int
	emotion    string
	format     string
	sampleRate int
	bitrate    int
}

var _ genx.Transformer = (*MinimaxTTS)(nil)

// MinimaxTTSOption is a functional option for MinimaxTTS.
type MinimaxTTSOption func(*MinimaxTTS)

// WithMinimaxTTSModel sets the model.
func WithMinimaxTTSModel(model string) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.model = model
	}
}

// WithMinimaxTTSSpeed sets the speech speed (0.5-2.0).
func WithMinimaxTTSSpeed(speed float64) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.speed = speed
	}
}

// WithMinimaxTTSVolume sets the volume (0-10).
func WithMinimaxTTSVolume(vol float64) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.vol = vol
	}
}

// WithMinimaxTTSPitch sets the pitch adjustment (-12 to 12).
func WithMinimaxTTSPitch(pitch int) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.pitch = pitch
	}
}

// WithMinimaxTTSEmotion sets the emotion.
// Options: happy, sad, angry, fearful, disgusted, surprised, neutral
func WithMinimaxTTSEmotion(emotion string) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.emotion = emotion
	}
}

// WithMinimaxTTSFormat sets the audio format.
// Options: mp3, pcm, flac, wav
func WithMinimaxTTSFormat(format string) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.format = format
	}
}

// WithMinimaxTTSSampleRate sets the sample rate.
// Options: 8000, 16000, 22050, 24000, 32000, 44100
func WithMinimaxTTSSampleRate(sampleRate int) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.sampleRate = sampleRate
	}
}

// WithMinimaxTTSBitrate sets the bitrate.
// Options: 32000, 64000, 128000, 256000
func WithMinimaxTTSBitrate(bitrate int) MinimaxTTSOption {
	return func(t *MinimaxTTS) {
		t.bitrate = bitrate
	}
}

// NewMinimaxTTS creates a new MinimaxTTS transformer.
//
// Parameters:
//   - client: MiniMax client
//   - voiceID: Voice identifier (e.g., "female-shaonv", "male-qn-qingse")
//   - opts: Optional configuration
func NewMinimaxTTS(client *minimax.Client, voiceID string, opts ...MinimaxTTSOption) *MinimaxTTS {
	t := &MinimaxTTS{
		client:     client,
		model:      "speech-02-hd",
		voiceID:    voiceID,
		speed:      1.0,
		vol:        1.0,
		format:     "mp3",
		sampleRate: 32000,
		bitrate:    128000,
	}
	for _, opt := range opts {
		opt(t)
	}
	return t
}

// MinimaxTTSCtxKey is the context key for runtime options.
type minimaxTTSCtxKey struct{}

// MinimaxTTSCtxOptions are runtime options passed via context.
// TODO: Add fields as needed for runtime configuration.
type MinimaxTTSCtxOptions struct{}

// WithMinimaxTTSCtxOptions attaches runtime options to context.
func WithMinimaxTTSCtxOptions(ctx context.Context, opts MinimaxTTSCtxOptions) context.Context {
	return context.WithValue(ctx, minimaxTTSCtxKey{}, opts)
}

// Transform converts Text chunks to audio Blob chunks.
// MinimaxTTS does not require connection setup, so it returns immediately.
// The ctx is unused (no initialization needed); the goroutine lifetime
// is governed by the input Stream.
func (t *MinimaxTTS) Transform(_ context.Context, _ string, input genx.Stream) (genx.Stream, error) {
	output := newBufferStream(100)

	go t.transformLoop(input, output)

	return output, nil
}

func (t *MinimaxTTS) transformLoop(input genx.Stream, output *bufferStream) {
	defer output.Close()

	// Local cancel context tied to the loop lifecycle.
	// When the loop exits, defer cancel() cancels any in-flight HTTP request.
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mimeType := t.mimeType()
	var textBuilder strings.Builder
	var lastChunk *genx.MessageChunk
	var currentStreamID string

	for {
		chunk, err := input.Next()
		if err != nil {
			if err != io.EOF {
				output.CloseWithError(err)
				return
			}
			// EOF: synthesize any remaining text
			if textBuilder.Len() > 0 {
				if err := t.synthesize(ctx, textBuilder.String(), lastChunk, currentStreamID, mimeType, output); err != nil {
					output.CloseWithError(err)
					return
				}
			}
			return
		}

		if chunk == nil {
			continue
		}

		lastChunk = chunk

		// Track StreamID from input - inherit or generate new one
		if chunk.Ctrl != nil && chunk.Ctrl.StreamID != "" {
			currentStreamID = chunk.Ctrl.StreamID
		} else if currentStreamID == "" {
			// Generate new StreamID if none provided
			currentStreamID = genx.NewStreamID()
		}

		// Check for text EoS marker
		if chunk.IsEndOfStream() {
			if _, ok := chunk.Part.(genx.Text); ok {
				// Text EoS: synthesize accumulated text, emit audio, then emit audio EoS
				if textBuilder.Len() > 0 {
				if err := t.synthesize(ctx, textBuilder.String(), lastChunk, currentStreamID, mimeType, output); err != nil {
					output.CloseWithError(err)
					return
				}
				textBuilder.Reset()
				}
				// Emit audio EoS with StreamID
				eosChunk := &genx.MessageChunk{
					Part: &genx.Blob{MIMEType: mimeType},
					Ctrl: &genx.StreamCtrl{StreamID: currentStreamID, EndOfStream: true},
				}
				if lastChunk != nil {
					eosChunk.Role = lastChunk.Role
					eosChunk.Name = lastChunk.Name
				}
				if err := output.Push(eosChunk); err != nil {
					return
				}
				// Reset StreamID for next synthesis segment
				currentStreamID = ""
				continue
			}
			// Non-text EoS: pass through with StreamID
			if chunk.Ctrl == nil {
				chunk.Ctrl = &genx.StreamCtrl{}
			}
			chunk.Ctrl.StreamID = currentStreamID
			if err := output.Push(chunk); err != nil {
				return
			}
			continue
		}

		// Collect text
		if text, ok := chunk.Part.(genx.Text); ok {
			textBuilder.WriteString(string(text))
		} else {
			// Non-text chunk: pass through with StreamID
			if chunk.Ctrl == nil {
				chunk.Ctrl = &genx.StreamCtrl{}
			}
			chunk.Ctrl.StreamID = currentStreamID
			if err := output.Push(chunk); err != nil {
				return
			}
		}
	}
}

func (t *MinimaxTTS) synthesize(ctx context.Context, text string, lastChunk *genx.MessageChunk, streamID, mimeType string, output *bufferStream) error {
	// Emit BOS at the start of synthesis
	bosChunk := &genx.MessageChunk{
		Ctrl: &genx.StreamCtrl{StreamID: streamID, BeginOfStream: true},
	}
	if lastChunk != nil {
		bosChunk.Role = lastChunk.Role
		bosChunk.Name = lastChunk.Name
	}
	if err := output.Push(bosChunk); err != nil {
		return err
	}

	req := &minimax.SpeechRequest{
		Model: t.model,
		Text:  text,
		VoiceSetting: &minimax.VoiceSetting{
			VoiceID: t.voiceID,
			Speed:   t.speed,
			Vol:     t.vol,
			Pitch:   t.pitch,
			Emotion: t.emotion,
		},
		AudioSetting: &minimax.AudioSetting{
			Format:     minimax.AudioFormat(t.format),
			SampleRate: t.sampleRate,
			Bitrate:    t.bitrate,
		},
	}

	for chunk, err := range t.client.Speech.SynthesizeStream(ctx, req) {
		if err != nil {
			return err
		}

		if chunk.Audio != nil && len(chunk.Audio) > 0 {
			outChunk := &genx.MessageChunk{
				Part: &genx.Blob{
					MIMEType: mimeType,
					Data:     chunk.Audio,
				},
				Ctrl: &genx.StreamCtrl{StreamID: streamID},
			}
			if lastChunk != nil {
				outChunk.Role = lastChunk.Role
				outChunk.Name = lastChunk.Name
			}

			if err := output.Push(outChunk); err != nil {
				return err
			}
		}
	}
	return nil
}

func (t *MinimaxTTS) mimeType() string {
	switch t.format {
	case "mp3":
		return "audio/mpeg"
	case "pcm":
		return "audio/pcm"
	case "flac":
		return "audio/flac"
	case "wav":
		return "audio/wav"
	default:
		return "audio/mpeg"
	}
}

File: rust/genx/src/transformers/minimax_tts.rs
use std::sync::Arc;

use async_trait::async_trait;
use futures::{pin_mut, StreamExt};
use giztoy_minimax as minimax;

use crate::error::GenxError;
use crate::stream::Stream;
use crate::transformer::Transformer;

use super::tts_core::{spawn_tts_transform_loop, AudioEmitter, TtsProvider};

pub struct MinimaxTtsTransformer {
    provider: Arc<dyn TtsProvider>,
}

impl MinimaxTtsTransformer {
    pub fn new(client: Arc<minimax::Client>, voice_id: impl Into<String>) -> Self {
        Self {
            provider: Arc::new(MinimaxProvider {
                client,
                model: "speech-2.6-hd".to_string(),
                voice_id: voice_id.into(),
                speed: 1.0,
                vol: 1.0,
                pitch: 0,
                emotion: None,
                format: "mp3".to_string(),
                sample_rate: Some(32000),
                bitrate: Some(128000),
            }),
        }
    }

    pub(crate) fn new_with_provider(provider: Arc<dyn TtsProvider>) -> Self {
        Self { provider }
    }
}

#[async_trait]
impl Transformer for MinimaxTtsTransformer {
    async fn transform(
        &self,
        _pattern: &str,
        input: Box<dyn Stream>,
    ) -> Result<Box<dyn Stream>, GenxError> {
        Ok(spawn_tts_transform_loop(Arc::clone(&self.provider), input))
    }
}

struct MinimaxProvider {
    client: Arc<minimax::Client>,
    model: String,
    voice_id: String,
    speed: f64,
    vol: f64,
    pitch: i32,
    emotion: Option<String>,
    format: String,
    sample_rate: Option<i32>,
    bitrate: Option<i32>,
}

#[async_trait]
impl TtsProvider for MinimaxProvider {
    fn mime_type(&self) -> &str {
        match self.format.as_str() {
            "mp3" => "audio/mpeg",
            "pcm" => "audio/pcm",
            "flac" => "audio/flac",
            "wav" => "audio/wav",
            _ => "audio/mpeg",
        }
    }

    async fn synthesize_stream(
        &self,
        text: &str,
        emitter: &mut dyn AudioEmitter,
    ) -> Result<(), GenxError> {
        let req = minimax::SpeechRequest {
            model: self.model.clone(),
            text: text.to_string(),
            voice_setting: Some(minimax::VoiceSetting {
                voice_id: self.voice_id.clone(),
                speed: Some(self.speed),
                vol: Some(self.vol),
                pitch: Some(self.pitch),
                emotion: self.emotion.clone(),
            }),
            audio_setting: Some(minimax::AudioSetting {
                sample_rate: self.sample_rate,
                bitrate: self.bitrate,
                format: Some(match self.format.as_str() {
                    "pcm" => minimax::AudioFormat::Pcm,
                    "flac" => minimax::AudioFormat::Flac,
                    "wav" => minimax::AudioFormat::Wav,
                    _ => minimax::AudioFormat::Mp3,
                }),
                channel: None,
            }),
            ..Default::default()
        };

        let speech = self.client.speech();
        let stream = speech
            .synthesize_stream(&req)
            .await
            .map_err(|e| GenxError::Other(anyhow::anyhow!("minimax tts open stream: {}", e)))?;

        pin_mut!(stream);
        while let Some(item) = stream.next().await {
            let chunk = item
                .map_err(|e| GenxError::Other(anyhow::anyhow!("minimax tts stream chunk: {}", e)))?;
            emitter.emit(chunk.audio).await?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{MessageChunk, Part, Role, StreamCtrl};
    use tokio::sync::Mutex;

    struct MockProvider {
        mime: String,
        responses: Mutex<Vec<Result<Vec<Vec<u8>>, GenxError>>>,
    }

    #[async_trait]
    impl TtsProvider for MockProvider {
        fn mime_type(&self) -> &str {
            &self.mime
        }

        async fn synthesize_stream(
            &self,
            _text: &str,
            emitter: &mut dyn AudioEmitter,
        ) -> Result<(), GenxError> {
            match self.responses.lock().await.remove(0) {
                Ok(chunks) => {
                    for audio in chunks {
                        emitter.emit(audio).await?;
                    }
                    Ok(())
                }
                Err(e) => Err(e),
            }
        }
    }

    fn input_stream(chunks: Vec<MessageChunk>) -> Box<dyn Stream> {
        let builder = crate::stream::StreamBuilder::with_tools(16, vec![]);
        builder.add(&chunks).expect("add chunks");
        builder.done(crate::error::Usage::default()).expect("done");
        Box::new(builder.stream())
    }

    #[tokio::test]
    async fn tts_emits_bos_audio_eos_on_text_eos() {
        let provider: Arc<dyn TtsProvider> = Arc::new(MockProvider {
            mime: "audio/mpeg".to_string(),
            responses: Mutex::new(vec![Ok(vec![vec![1, 2], vec![3]])]),
        });
        let t = MinimaxTtsTransformer::new_with_provider(provider);

        let mut in_chunk = MessageChunk::text(Role::Model, "hello");
        in_chunk.name = Some("assistant".to_string());
        in_chunk.ctrl = Some(StreamCtrl {
            stream_id: "s1".to_string(),
            ..Default::default()
        });
        let eos = MessageChunk::new_text_end_of_stream();

        let mut out = t
            .transform("minimax/shaonv", input_stream(vec![in_chunk, eos]))
            .await
            .expect("transform");

        let c1 = out.next().await.expect("next1").expect("chunk1");
        assert!(c1.is_begin_of_stream());
        assert_eq!(c1.ctrl.as_ref().map(|c| c.stream_id.as_str()), Some("s1"));
        assert_eq!(c1.role, Role::Model);

        let c2 = out.next().await.expect("next2").expect("chunk2");
        assert!(matches!(c2.part, Some(Part::Blob(_))));
        assert_eq!(c2.ctrl.as_ref().map(|c| c.stream_id.as_str()), Some("s1"));

        let c3 = out.next().await.expect("next3").expect("chunk3");
        assert!(matches!(c3.part, Some(Part::Blob(_))));

        let c4 = out.next().await.expect("next4").expect("chunk4");
        assert!(c4.is_end_of_stream());
        assert_eq!(c4.ctrl.as_ref().map(|c| c.stream_id.as_str()), Some("s1"));
        assert!(out.next().await.expect("eof").is_none());
    }

    #[tokio::test]
    async fn tts_propagates_provider_error() {
        let provider: Arc<dyn TtsProvider> = Arc::new(MockProvider {
            mime: "audio/mpeg".to_string(),
            responses: Mutex::new(vec![Err(GenxError::Other(anyhow::anyhow!(
                "mock synth fail"
            )))]),
        });
        let t = MinimaxTtsTransformer::new_with_provider(provider);

        let mut out = t
            .transform(
                "minimax/shaonv",
                input_stream(vec![
                    MessageChunk::text(Role::Model, "hello"),
                    MessageChunk::new_text_end_of_stream(),
                ]),
            )
            .await
            .expect("transform");

        let _ = out.next().await.expect("bos").expect("bos chunk");
        let err = out.next().await.expect_err("expect provider error");
        assert!(err.to_string().contains("mock synth fail"));
    }
}
